### PR TITLE
AM-2911:  Judicial Refresh (via ASB or Job) to include Judicial Bookings

### PR DESCRIPTION
**Pull request by prashanthkumarnb**
### JIRA link (if applicable) ###

   [AM-2911](https://tools.hmcts.net/jira/browse/AM-2911) _"Judicial Refresh (via ASB or Job) does not include Judicial Bookings when generating organisation Role-Assignments"_

### Change description ###

Change is to include JBS lookup in the ASB Judicial Refresh and the Judicial Refresh Jobs.   **However, the change is behind env-var flags: which are disabled in this PR**.  To see a PR with the fllags enabled since the test PRs listed below. 

This is dependent on JBS PR 
* hmcts/am-judicial-booking-service#854

NB: This is the primary PR for this ticket:  the following PRs are for testing additional configuration scenarios:

* #1827
* #1832

> **NB: This PR now also contains the following additional changes: which are to be tested in parrallel**: 
> * #1826
> * #1889
> * #1869

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by mattnayler — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestrator.java [hunk 1]
```diff
@@ -53,16 +56,29 @@ public ResponseEntity<Object> createBulkAssignmentsRequest(UserRequest userReque
...
+        }
 
         log.debug("Execution time of createBulkAssignmentsRequest() : {} ms",
                 (Math.subtractExact(System.currentTimeMillis(), startTime)));
 
         List<Object> roleAssignmentResponses = new ArrayList<>();
-        ((List<ResponseEntity<Object>>)
-                Objects.requireNonNull(responseEntity.getBody())).forEach(entity ->
-                roleAssignmentResponses.add(entity.getBody()));
+        if (null != responseEntity) {
```
[the remark this answers is not in the document]
Comment (prashanthkumarnb) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestrator.java:77]: [Null_pointers_shouldNot_Dereferenced.pdf](https://github.com/hmcts/am-org-role-mapping-service/files/13120283/Null_pointers_shouldNot_Dereferenced.pdf)

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/JudicialBookingServiceTest.java [hunk 2]
```diff
@@ -68,4 +68,33 @@ void fetchJudicialBookings_emptyResponse() {
...
+    void testFetchJudicialBookingsInBatchesEmptyUsers() {
+        List<String> userIds = Collections.emptyList();
+        String batchSize = "400";
+
+        List<JudicialBooking> judBookings = sut.fetchJudicialBookingsInBatches(userIds,batchSize);
+        assertEquals(0,judBookings.size());
+    }
+
+    @Test
+    void testFetchJudicialBookingsInBatchesWithUsers() {
```
[the remark this answers is not in the document]
Comment (prashanthkumarnb) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/JudicialBookingServiceTest.java:82]: addressed

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestrator.java [hunk 3]
```diff
@@ -53,16 +56,29 @@ public ResponseEntity<Object> createBulkAssignmentsRequest(UserRequest userReque
...
+        if (userType.equals(UserType.JUDICIAL)) {
+            List<JudicialBooking> judicialBookings = judicialBookingService.fetchJudicialBookings(userRequest);
+            log.info("Judicial User {} profile(s) got {} booking(s)", userAccessProfiles.size(),
+                    judicialBookings.size());
+            responseEntity = requestMappingService.createJudicialAssignments(userAccessProfiles, judicialBookings);
+        } else if (userType.equals(UserType.CASEWORKER)) {
+            log.info("Creating assignments for Staff/Caseworker UserType");
+            responseEntity = requestMappingService.createCaseworkerAssignments(userAccessProfiles);
+        } else {
+            log.info("UserType not supported {}",userType);
```
[the remark this answers is not in the document]
Comment (prashanthkumarnb) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestrator.java]: addressed

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestrator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (prashanthkumarnb) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestrator.java:75]: addressed

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java [hunk 4]
```diff
@@ -57,18 +59,53 @@ public void setUp() {
     @SuppressWarnings("unchecked")
     void createBulkAssignmentsRequestTest() {
 
+        doReturn(TestDataBuilder.buildUserAccessProfileMap(false, false)).when(retrieveDataService)
```
[the remark this answers is not in the document]
Comment (prashanthkumarnb) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java:62]: addressed

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java [hunk 5]
```diff
@@ -101,7 +138,7 @@ void createBulkAssignmentsRequestForJudicial() {
         List<ResponseEntity<Object>> responseEntities = List.of(ResponseEntity.ok(AssignmentRequestBuilder
                 .buildJudicialAssignmentRequest(false)));
 
-        Mockito.when(requestMappingService.createAssignments(Mockito.any(), Mockito.any()))
+        Mockito.when(requestMappingService.createJudicialAssignments(Mockito.any(), Mockito.any()))
```
[the remark this answers is not in the document]
Comment (prashanthkumarnb) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java]: addressed

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java [hunk 6]
```diff
@@ -125,7 +162,7 @@ void createBulkAssignmentsRequestForJudicial() {
         Mockito.verify(retrieveDataService, Mockito.times(1))
                 .retrieveProfiles(Mockito.any(UserRequest.class),Mockito.any());
         Mockito.verify(requestMappingService, Mockito.times(1))
-                .createAssignments(Mockito.any(),Mockito.any());
+                .createJudicialAssignments(Mockito.any(),Mockito.any());
```
[the remark this answers is not in the document]
Comment (prashanthkumarnb) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java]: addressed

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/RefreshOrchestrator.java [hunk 7]
```diff
@@ -212,10 +212,30 @@ protected ResponseEntity<Object> refreshJobByServiceName(Map<String, HttpStatus>
...
+            //extract unique userids from userAccessProfiles
+            List<String> userIds = new ArrayList<>();
+
+            for (Set<UserAccessProfile> userAccessProfileSet:userAccessProfiles.values()) {
+                List<UserAccessProfile> userAccessProfileList = userAccessProfileSet.stream().toList();
+                for (UserAccessProfile userAccessProfile:userAccessProfileList) {
+                    userIds.add(((JudicialAccessProfile)userAccessProfile).getUserId());
+                }
+            }
+            List<String> uniqueUserIds = userIds.stream().distinct().toList();
```
[the remark this answers is not in the document]
Comment (prashanthkumarnb) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/RefreshOrchestrator.java:230]: the code looked complex with less readability , let me know when we pair we can discuss

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/RefreshOrchestratorTest.java [hunk 8]
```diff
@@ -619,22 +622,20 @@ void refreshRoleAssignmentRecords_judicial() {
...
-                .suspended(false)
                 .build());
+
         userAccessProfiles.put("1", userAccessProfileSet);
         RefreshOrchestrator refreshOrchestrator = Mockito.spy(sut);
         Mockito.when(retrieveDataService.retrieveProfiles(any(), eq(UserType.JUDICIAL)))
                 .thenReturn(userAccessProfiles);
 
-        Mockito.when(requestMappingService.createAssignments(any(), eq(UserType.JUDICIAL)))
+        Mockito.when(requestMappingService.createJudicialAssignments(any(), any()))
```
[the remark this answers is not in the document]
Comment (prashanthkumarnb) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/RefreshOrchestratorTest.java:641]: addressed

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/JudicialBookingService.java [hunk 9]
```diff
@@ -27,4 +32,24 @@ public List<JudicialBooking> fetchJudicialBookings(UserRequest userRequest) {
 
         return Objects.requireNonNullElse(response, JudicialBookingResponse.builder().build()).getBookings();
     }
+
+    //New service takes List of USERIDs and batchSize configured in refresh.job.pageSize
+    public List<JudicialBooking> fetchJudicialBookingsInBatches(List<String> userIds,String batchSize) {
+        log.info(" fetching Judicial Bookings for userIds {} and the batchSize is {}", userIds,batchSize);
+        List<JudicialBooking> judicialBookings = new ArrayList<>();
+        final int defaultBatchSize = Integer.parseInt(batchSize);
```
[the remark this answers is not in the document]
Comment (prashanthkumarnb) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/JudicialBookingService.java]: addressed

File: Jenkinsfile_CNP [hunk 10]
```diff
@@ -76,7 +76,8 @@ env.AMQP_HOST = "rd-servicebus-aat.servicebus.windows.net"
 env.CRD_TOPIC_NAME = "rd-caseworker-topic-aat"
 env.JRD_TOPIC_NAME = "rd-judicial-topic-aat"
 env.ROLE_ASSIGNMENT_URL = "http://am-role-assignment-service-aat.service.core-compute-aat.internal"
-env.JUDICIAL_BOOKING_URL = "http://am-judicial-booking-service-aat.service.core-compute-aat.internal"
+env.JUDICIAL_BOOKING_URL = "https://am-judicial-booking-service-pr-854.preview.platform.hmcts.net"
+// TODO env.JUDICIAL_BOOKING_URL = "http://am-judicial-booking-service-aat.service.core-compute-aat.internal"
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply Jenkinsfile_CNP]: Resolved in 457da23aac8e8b48f48d4648caafb65985e72af0.

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestrator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestrator.java]: Resolved in 9d50d3f80627384371507e4fce692d5c5a687968.

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/JudicialBookingServiceTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/JudicialBookingServiceTest.java:82]: Resolved in 9d50d3f80627384371507e4fce692d5c5a687968.

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java:64]: Resolved in 9d50d3f80627384371507e4fce692d5c5a687968.

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java]: Resolved in 9d50d3f80627384371507e4fce692d5c5a687968.

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestrator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestrator.java:75]: Also refactored in #1820.

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java [hunk 11]
```diff
@@ -57,18 +59,53 @@ public void setUp() {
...
+                .validateUserRequest(Mockito.any(UserRequest.class));
+        Mockito.verify(retrieveDataService, Mockito.times(1))
+                .retrieveProfiles(Mockito.any(UserRequest.class),Mockito.any());
+        Mockito.verify(requestMappingService, Mockito.times(1))
+                .createCaseworkerAssignments(Mockito.any());
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    void createJudicialBulkAssignmentsRequestTest() {
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/BulkAssignmentOrchestratorTest.java:99]: Also refactored in #1820.

File: src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/controller/WelcomeControllerIntegrationTest.java [hunk 12]
```diff
@@ -565,26 +553,26 @@ public void createOrgRoleMappingDroolsMustMapCorrectRoleName() throws Exception
     @DisplayName("S20: drools must map correct roles name based on appointments & appointment type")
     public void createOrgRolesForIACTribunalJudge_FeePaidThroughMapping() throws Exception {
 
-
         judicialProfiles.get(0).getAppointments().remove(0);
 
         doReturn(ResponseEntity.ok(judicialProfiles))
                 .when(jrdFeignClient).getJudicialDetailsById(any(), any());
 
+        doReturn(buildJudicialBookingsResponse()).when(jbsFeignClient).getJudicialBookingByUserIds(any());
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/controller/WelcomeControllerIntegrationTest.java:567]: Refactored in #1820.

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/RefreshOrchestrator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/RefreshOrchestrator.java:230]: Refactored in #1820.  Code moved to utility class to simplify edge case testing.

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/RefreshOrchestratorTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/RefreshOrchestratorTest.java:641]: Refactored in #1820.  The code for conveting the nested map in to a list has been moved to utility class to simplify edge case testing.

File: src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/controller/WelcomeControllerIntegrationTest.java [hunk 13]
```diff
@@ -137,9 +142,12 @@ public class WelcomeControllerIntegrationTest extends BaseTest {
     private static final String RAS_DELETE_FLAG_TRUE = "RASDeleteFlagTrue";
     private static final String RAS_DROOL_RULE_FAIL = "RASDroolRuleFail";
     private static final String RAS_UPDATE_ROLE_TCW_STCW = "RASUpdateRoleTcwStcw";
-
     private static final String ROLE_NAME_STCW = "senior-tribunal-caseworker";
     private static final String ROLE_NAME_TCW = "tribunal-caseworker";
+    private static final String ROLE_NAME_HMCTS_JUD = "hmcts-judiciary";
+    private static final String USERID_74 = "123e4567-e89b-42d3-a456-556642445674";
+    private static final String USERID_9c = "21334a2b-79ce-44eb-9168-2d49a744be9c";
+    private static final String  STAFF_URI = "/am/role-mapping/staff/users";
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/controller/WelcomeControllerIntegrationTest.java:152]: Refactored in #1820.

**Review by mattnayler — DISMISSED** [review]
Re-reviewed after merger of #1820 (75bbff07bef5dc5583db5732fc93dffc3e880b7a) and master (5c2e3a048b6d22d174718c4f426d861b3cd89926).

**Review by mikebrownccd — DISMISSED** [review]
LGTM

**Review by mattnayler — APPROVED** [review]
Re-approved after merger of: 
* #1826
* #1889
* #1869

**Review by mikebrownccd — APPROVED** [review]
LGTM
